Can not control virtual time for multiple delay calls when using runTest

When using `0.12.0`, it appears that the virtual time of the test cannot be controlled when using `runTest` after the first call to `awaitItem` when there are multiple `delay` calls. Consider the following test:

```kotlin
    @Test
    fun virtualTimeCanBeControlled() = runTest {
        flow {
            delay(5000)
            emit("1")
            delay(5000)
            emit("2")
        }
            .test {
                expectNoEvents()

                advanceTimeBy(5000)

                expectNoEvents()
                runCurrent()
                assertEquals("1", awaitItem())

                advanceTimeBy(5000)

                expectNoEvents() // <---------- Fails here
                runCurrent()
                assertEquals("2", awaitItem())

                awaitComplete()
            }
    }
```

This fails with the following error where indicated:

```
Expected no events but found Item(2)
app.cash.turbine.TurbineAssertionError: Expected no events but found Item(2)
	at app//app.cash.turbine.ChannelKt.unexpectedEvent(channel.kt:258)
	at app//app.cash.turbine.ChannelKt.unexpectedResult-JslgfBc(channel.kt:253)
	at app//app.cash.turbine.ChannelKt.expectNoEvents(channel.kt:71)
	at app//app.cash.turbine.ChannelTurbine.expectNoEvents(Turbine.kt:184)
	...
```

This test passes, however, when using `runBlocking` and a manually-supplied `UnconfinedTestDispatcher`:

```kotlin
    private val testDispatcher = UnconfinedTestDispatcher()

    @Test
    fun virtualTimeCanBeControlled() = runBlocking {
        flow {
            delay(5000)
            emit("1")
            delay(5000)
            emit("2")
        }
            .flowOn(testDispatcher)
            .test {
                expectNoEvents()

                testDispatcher.scheduler.advanceTimeBy(5000)

                expectNoEvents()
                testDispatcher.scheduler.runCurrent()
                assertEquals("1", awaitItem())

                testDispatcher.scheduler.advanceTimeBy(5000)

                expectNoEvents()
                testDispatcher.scheduler.runCurrent()
                assertEquals("2", awaitItem())

                awaitComplete()
            }
    }
```

This may be related to the changes in https://github.com/cashapp/turbine/pull/151, which bases logic on whether the current scope is using a test scheduler.

- Kotlin version : `1.7.10`
- Coroutines version : `1.6.4`

--- a/build.gradle
+++ b/build.gradle
@@ -1,9 +1,9 @@
 buildscript {
   dependencies {
-    classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.10'
+    classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.20'
     classpath 'com.diffplug.spotless:spotless-plugin-gradle:5.15.1'
     classpath 'com.vanniktech:gradle-maven-publish-plugin:0.14.2'
-    classpath 'org.jetbrains.dokka:dokka-gradle-plugin:1.7.10'
+    classpath 'org.jetbrains.dokka:dokka-gradle-plugin:1.7.20'
   }
 
   ext.versions = [

--- a/src/commonMain/kotlin/app/cash/turbine/channel.kt
+++ b/src/commonMain/kotlin/app/cash/turbine/channel.kt
@@ -20,8 +20,10 @@ import kotlin.time.Duration
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.DelicateCoroutinesApi
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.TimeoutCancellationException
 import kotlinx.coroutines.async
 import kotlinx.coroutines.channels.ChannelResult
@@ -115,7 +117,11 @@ private suspend fun <T> withWallclockTimeout(
   block: suspend CoroutineScope.() -> T,
 ): T = coroutineScope {
   val blockDeferred = async(start = CoroutineStart.UNDISPATCHED, block = block)
-  val timeoutJob = launch(Dispatchers.Default) { delay(timeout) }
+
+  // Run the timeout on a scope separate from the caller. This ensures that the use of the
+  // Default dispatcher does not affect the use of a TestScheduler and its fake time.
+  @OptIn(DelicateCoroutinesApi::class)
+  val timeoutJob = GlobalScope.launch(Dispatchers.Default) { delay(timeout) }
 
   select {
     blockDeferred.onAwait { result ->

--- a/src/commonTest/kotlin/app/cash/turbine/FlowTest.kt
+++ b/src/commonTest/kotlin/app/cash/turbine/FlowTest.kt
@@ -44,6 +44,8 @@ import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.onCompletion
 import kotlinx.coroutines.flow.onStart
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.advanceTimeBy
+import kotlinx.coroutines.test.runCurrent
 import kotlinx.coroutines.test.runTest
 import kotlinx.coroutines.withContext
 import kotlinx.coroutines.yield
@@ -654,6 +656,32 @@ class FlowTest {
         skipItems(3)
       }.message
       assertEquals("Expected 3 items for two item channel but got 2 items and Complete", message)
+    }
+  }
+
+  @Test
+  fun virtualTimeCanBeControlled() = runTest {
+    flow {
+      delay(5000)
+      emit("1")
+      delay(5000)
+      emit("2")
+    }.test {
+      expectNoEvents()
+
+      advanceTimeBy(5000)
+      expectNoEvents()
+
+      runCurrent()
+      assertEquals("1", awaitItem())
+
+      advanceTimeBy(5000)
+      expectNoEvents()
+
+      runCurrent()
+      assertEquals("2", awaitItem())
+
+      awaitComplete()
     }
   }
 }

